src/oneapi/oneapi.h

```diff
@@ -67,72 +68,201 @@ static std::string to_std_string(PyObject * o)
     return PyUnicode_AsUTF8(o);
 }
 
+#if INTEL_DAAL_VERSION >= 20210200
+inline const sycl::queue& get_current_queue()
+{
+    auto& ctx = daal::services::Environment::getInstance()->getDefaultExecutionContext();
+    auto* sycl_ctx = dynamic_cast<daal::services::internal::sycl::SyclExecutionContextImpl*>(&ctx);
+    if (!sycl_ctx)
+    {
+        throw std::domain_error("Cannot get current queue outside sycl_context");
+    }
+    return sycl_ctx->getQueue();
+}
+
+// take a raw array and convert to usm pointer
+template <typename T>
+inline daal::services::SharedPtr<T>* to_usm(T * ptr, int * shape)
+{
+    auto queue = get_current_queue();
+    const std::int64_t count = shape[0] * shape[1];
+    T* usm_host_ptr = sycl::malloc_host<T>(count, queue);
+    T* usm_device_ptr = sycl::malloc_device<T>(count, queue);
+    if (!usm_host_ptr || !usm_device_ptr)
+    {
+        sycl::free(usm_host_ptr, queue);
+        sycl::free(usm_device_ptr, queue);
+        throw std::runtime_error("internal error during allocating USM memory");
+    }
+
+    // TODO: avoid using usm_host_ptr and copy directly to usm_device_ptr
+    // It's a temporary solution till queue.memcpy() from non-usm memory does not work
+    int res = daal::services::internal::daal_memcpy_s(usm_host_ptr, sizeof(T) * count, ptr, sizeof(T) * count);
+    if (res)
+    {
+        sycl::free(usm_host_ptr, queue);
+        sycl::free(usm_device_ptr, queue);

```

Comment:
Shouldn't we wrap `free` calls into `try{...}catch(...){...}`?

Reply:
No, they do not throw exceptions